File: algorithms/eulerclj/problem15.clj

(defn paths-at [i grid cols]
  (let [left (if (= 0 (rem i cols)) 0 (nth grid (dec i)))
        up (if (< i cols) 0 (nth grid (- i cols)))]
    (+ left up)))

(defn lattice-paths [n]
  (let [cols (inc n)
        size (int (Math/pow cols 2))]
    (loop [path-grid [1] i 1]
      (let [paths-at-i (paths-at i path-grid cols)] 
        (if (= i size)
          (last path-grid)
          (recur (conj path-grid paths-at-i) (inc i)))))))


(time (println (lattice-paths 20)))

File: algorithms/eulerclj/problem15.rb

def lattice_paths(n)
  cols = n + 1
  size = (cols**2)
  grid = [1]

  1.upto(size-1) do |i|
    paths_left = ((i % cols) == 0) ? 0 : grid[i-1]
    paths_up = i < cols ? 0 : grid[i - cols]

    grid << paths_left + paths_up
  end

  return grid.last
end


puts lattice_paths(20)
